### PR TITLE
First attempt at fully qualifying the default values of C# properties.

### LINKED ISSUE #65938: Exporting nums in a separate namespace throw an error
### Godot version

4.0beta1

### System information

Windows 10

### Issue description

Fails to compile, receive this error in the build debugger:
`C:\...\Godot 4 Tests\Godot.SourceGenerators\Godot.SourceGenerators.ScriptPropertyDefValGenerator\NodeExt_ScriptPropertyDefVal_Generated.cs(11,69): The name 'EnumInAnotherNamespace' does not exist in the current context`

### Steps to reproduce

Drop the provided .cs file into your project and attempt to compile

Alternatively, attempt to [Export] any enum variable whose type is defined in a namespace different from the exporting namespace.

### Minimal reproduction project

[NodeExt.zip](https://github.com/godotengine/godot/files/9587250/NodeExt.zip)

The script attached above is literally just this:

```
using Godot;
using MyNamespace;

public partial class NodeExt : Node
{
    [Export] EnumInAnotherNamespace myMode = EnumInAnotherNamespace.PlsExportThis;
}

namespace MyNamespace
{
    public enum EnumInAnotherNamespace { Testing, PlsExportThis, sadface }
}

```

### LINKED ISSUE #66523: C# : Source generator fails to recognize default exported properties that reference other things in the assembly if their name isn't fully qualified.
### Godot version

4.0 beta 1

### System information

Windows 10 Home, GLES3, Intel Core i7-6567U, Intel Iris Graphics 550 (30.0.100.9955), 16GB RAM

### Issue description

If the default value of an exported property references a function/enum value/property that doesn't have a fully qualified name, building the project fails because the source generator is unable to recognize that symbol.

### Steps to reproduce

Make a node with an exported property with a default value that references a symbol that isn't part of the same namespace as the current class that has that namespace included with a using statement. The source generator spits out an error and building the project fails.

### Minimal reproduction project

[Godot4SourceGeneratorBug.zip](https://github.com/godotengine/godot/files/9660939/Godot4SourceGeneratorBug.zip)

### PR DESCRIPTION
**Pull request by rhofour**
I believe this fixes #66523.

This is my first time working with C# source generation so it's very possible I've missed something. I'm not particularly happy 
with how the spacing looks after SyntaxToQualifiedString, but it compiles fine.

I think this handles the issue better than #65907, but my PR doesn't handle the "this." part.

*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot/issues/65938.
- Fixes https://github.com/godotengine/godot/issues/66523.

### CODE REVIEW

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:383]: Why is adding a space needed? This seems to add unnecessary spaces to the generated syntax.

**Review by raulsntos — CHANGES_REQUESTED** [review]
Thanks for contributing! I personally prefer this approach over https://github.com/godotengine/godot/pull/65907 but I'm worried it won't work well with more complex default values such as:

```csharp
[Export]
public int[] MyArrayFromList = new List<int>(Array.Empty<int>()).ToArray();
```

Which doesn't work with your current PR, right now it generates:

```csharp
new List < int > ( System.Array . Empty < int > ( ) ) . ToArray ( );
```

Also, when you are done make sure to squash all commits into one and write a meaningful commit message such as `Fully qualify default values of C# properties`. Take a look at [CONTRIBUTING.md](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#contributing-pull-requests) and the [PR Workflow docs](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html) if you haven't already for more information, but don't hesitate to ask if you have any questions. You can also reach me on [Rocket Chat](https://chat.godotengine.org).

**Comment by rhofour** [comment]
Thanks for the review!

I don't see what's wrong with:
`new List < int > ( System.Array . Empty < int > ( ) ) . ToArray ( );`
As far as I can tell that will compile just fine.

**Comment by neikeq** [comment]
> Thanks for the review!
> 
> I don't see what's wrong with: `new List < int > ( System.Array . Empty < int > ( ) ) . ToArray ( );` As far as I can tell that will compile just fine.

I think he means that should be `new System.Collections.Generic.List < int > ( ...` instead.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:383]: Because otherwise we'd end up with no spaces at all in the output (which would then obviously fail to compile). I'd love to avoid the unnecessary spaces, but I haven't been able to figure out a good rule for when to add them and when not.

Comment (raulsntos) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:383]: You should be able to just use the existing whitespace from the nodes and tokens, it should be in the leading and trailing trivia.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:182]: I like it as an extension method. Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:369]: Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:383]: Ah, that does work, thanks!

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:372]: Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:363]: Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:376]: I can do this, but then the compile can't understand that AsNode won't be null. I slightly prefer this since it doesn't trigger any warnings, but I can change it if you'd prefer the other way.

**Comment by rhofour** [comment]
I've addressed almost all the comments locally and I've expanded it so it can handle cases like:
 `public int[] MyArrayFromList = new List<int>(Array.Empty<int>()).ToArray();`

I should be able to remove the foreach loop and post a new commit on Sunday.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs]: I did that, but I had to raise the target framework from netstandard2.0 to net6.0 to get access to DistinctBy. I believe this is the default for C# projects so I don't think this should be an issue, but if need be we can always implement DistinctBy ourselves and go back to netstandard2.0.

**Comment by rhofour** [comment]
I think I've addressed all the comments. Please verify everything looks good, then I'll squash the commits together.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (raulsntos) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs]: Oh, I didn't realize `DistinctBy` is .NET 6.0+ only. We can't raise the target to `net6.0` though because source generators must target `netstandard2.0`.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs
Comment (raulsntos) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs]: I see, I suppose we could use the null forgiveness operator (`!`).

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:184]: Our coding style uses new lines even for single lines:

```suggestion
                if (child.HasLeadingTrivia)
                {
                    sb.Append(child.GetLeadingTrivia());
                }
```

Although we sometimes omit the braces, I think that's what @neikeq would prefer. I tend to prefer to always add braces but I don't have a strong opinion.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:195]: Is trimming the end here really necessary?

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:188]: Do you think you could rewrite the `FullQualifiedSyntax` method as a private method that takes a `StringBuilder` and appends to it instead of returning a `string` to avoid allocating multiple strings? Then the public method would just create the `StringBuilder` and pass it to the private method.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:180]: Why use a capacity of 128?

Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:180]: It's just an hint, but 128 is probably longer than we'd usually want. Would you prefer it at something smaller like 64 or would you rather we just not provide a size hint at all?

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:188]: Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs:195]: No, it used to be when we were adding the spaces, but it doesn't make sense anymore. Removed.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:173]: Ideally we would do this just once per syntax tree. I think the best way to do this is to go back to the foreach loop I had before where we processed all the classes in a syntax tree at once.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs]: Ah, I'll put that back then. I think going back to the loop also lets me avoid re-generating the SemanticModel anyways.

**Comment by rhofour** [comment]
I think my latest commit addresses the most recent comments.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs
Comment (raulsntos) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs]: It's not just a hint, the capacity determines the size of the internal array that will be allocated so I'd rather not allocate a bigger array than we need. It's probably better to use the parameterless constructor so it uses the default capacity, I'm sure the default value is likely to work best in most situations which is why it was chosen.

Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ExtensionMethods.cs]: Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (neikeq) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:24]: I can't think properly right now, but this part doesn't seem right. You're iterating over all syntax trees looking for godot classes, but you only process the properties for the first syntax tree that finds a godot class. This syntax tree may not be the one containing a property/field initializer we need. In fact, a class could be divided into multiple syntax trees (with `partial`), each one declaring properties with initializers.

If all of this is to get the semantic model, you could get it inline from the property initializer. See my other review comment.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (neikeq) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:178]: Following my other review comment, the semantic model could be obtained here:

```suggestion
                string? value;

                if (initializer != null)
                {
                    var sm = context.Compilation.GetSemanticModel(initializer.SyntaxTree);

                    // Fully qualify the value to avoid issues with namespaces.
                    value = initializer.Value.FullQualifiedSyntax(sm);
                }
                else
                {
                    value = null;
                }
```

It can be extracted to a method if it's too verbose.

Comment (neikeq) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs]: I don't know how much it costs to get the semantic model. If it ends up being a problem, a way to reduce repeated calls could be to iterate by syntax trees of the class symbol locations rather than by properties. In any case, I wouldn't worry about that in this PR.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (neikeq) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:174]:
```suggestion
```

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (neikeq) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:218]:
```suggestion
```

**Review by neikeq — CHANGES_REQUESTED** [review]
PR looks good to merge but two old lines were left commented.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs]: Done.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
Comment (rhofour) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs]: Done.

**Comment by rhofour** [comment]
Now that @neikeq said it looks good to merge I've squashed everything (after removing the two old, commented out lines).

**Review by neikeq — APPROVED** [review]
Looks good. I wouldn't be surprised if we discover other unhandled scenarios in the future (like the one with `new List`), but I prefer this approach over other alternatives.

**Comment by rhofour** [comment]
Ok, it looks like I can't merge it with the outstanding change requests from @raulsntos. Once that's resolved I'll merge it (or one of you can).

**Review by raulsntos — APPROVED** [review]
Left a nitpick but looks good to merge regardless.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
